File: src/solar/solar-common/w3ts/handles/effect.ts
/** @noSelfInFile **/

import {Handle} from "./handle";
import {Widget} from "./widget";

export class Effect extends Handle<effect> {

    /**特效对象 */
    private _data: any

    get data(): any {
        if (!this._data) {
            this._data = {}
        }
        return this._data;
    }

    set data(value: any) {
        this._data = value;
    }

    constructor(modelName: string, x: number, y: number);
    constructor(modelName: string, targetWidget: Widget, attachPointName: string);
    constructor(modelName: string, a: number | Widget, b: number | string) {
        if (Handle.initFromHandle()) {
            super();
        } else if (typeof a === "number" && typeof b === "number") {
            super(AddSpecialEffect(modelName, a, b));
        } else if (typeof a !== "number" && typeof b === "string") {
            super(AddSpecialEffectTarget(modelName, a.handle, b));
        }
    }

    // public get scale() {
    //   return BlzGetSpecialEffectScale(this.handle);
    // }
    //
    // public set scale(scale: number) {
    //   BlzSetSpecialEffectScale(this.handle, scale);
    // }

    /**
     * Warning: asynchronous
     */
    // public get x() {
    //   return BlzGetLocalSpecialEffectX(this.handle);
    // }
    //
    // public set x(x: number) {
    //   BlzSetSpecialEffectX(this.handle, x);
    // }

    /**
     * Warning: asynchronous
     */
    // public get y() {
    //   return BlzGetLocalSpecialEffectY(this.handle);
    // }
    //
    // public set y(y: number) {
    //   BlzSetSpecialEffectY(this.handle, y);
    // }
    //
    // /**
    //  * Warning: asynchronous
    //  */
    // public get z() {
    //   return BlzGetLocalSpecialEffectZ(this.handle);
    // }
    //
    // public set z(z: number) {
    //   BlzSetSpecialEffectZ(this.handle, z);
    // }
    //
    // public addSubAnimation(subAnim: subanimtype) {
    //   BlzSpecialEffectAddSubAnimation(this.handle, subAnim);
    // }
    //
    // public clearSubAnimations() {
    //   BlzSpecialEffectClearSubAnimations(this.handle);
    // }

    // public playAnimation(animType: animtype) {
    //   BlzPlaySpecialEffect(this.handle, animType);
    // }
    //
    // public playWithTimeScale(animType: animtype, timeScale: number) {
    //   BlzPlaySpecialEffectWithTimeScale(this.handle, animType, timeScale);
    // }
    //
    // public removeSubAnimation(subAnim: subanimtype) {
    //   BlzSpecialEffectRemoveSubAnimation(this.handle, subAnim);
    // }
    //
    // public resetScaleMatrix() {
    //   BlzResetSpecialEffectMatrix(this.handle);
    // }
    //
    // public setAlpha(alpha: number) {
    //   BlzSetSpecialEffectAlpha(this.handle, alpha);
    // }
    //
    // public setColor(red: number, green: number, blue: number) {
    //   BlzSetSpecialEffectColor(this.handle, red, green, blue);
    // }
    //
    // public setColorByPlayer(whichPlayer: MapPlayer) {
    //   BlzSetSpecialEffectColorByPlayer(this.handle, whichPlayer.handle);
    // }
    //
    // public setHeight(height: number) {
    //   BlzSetSpecialEffectHeight(this.handle, height);
    // }
    //
    // public setOrientation(yaw: number, pitch: number, roll: number) {
    //   BlzSetSpecialEffectOrientation(this.handle, yaw, pitch, roll);
    // }

    // public setPitch(pitch: number) {
    //   BlzSetSpecialEffectPitch(this.handle, pitch);
    // }
    //
    // public setPoint(p: Point) {
    //   BlzSetSpecialEffectPositionLoc(this.handle, p.handle);
    // }
    //
    // public setPosition(x: number, y: number, z: number) {
    //   BlzSetSpecialEffectPosition(this.handle, x, y, z);
    // }
    //
    // public setRoll(roll: number) {
    //   BlzSetSpecialEffectRoll(this.handle, roll);
    // }
    //
    // public setScaleMatrix(x: number, y: number, z: number) {
    //   BlzSetSpecialEffectMatrixScale(this.handle, x, y, z);
    // }
    //
    // public setTime(value: number) {
    //   BlzSetSpecialEffectTime(this.handle, value);
    // }
    //
    // public setTimeScale(timeScale: number) {
    //   BlzSetSpecialEffectTimeScale(this.handle, timeScale);
    // }

    // public setYaw(y: number) {
    //   BlzSetSpecialEffectYaw(this.handle, y);
    // }

    /**
     * 新建特效(创建到坐标)
     */
    public static createEffect(modelName: string, x: number, y: number) {
        return new Effect(modelName, x, y);
    }

    /**
     * 新建特效(绑定到单位)
     */
    public static createEffectTarget(modelName: string, a: Widget, b: string) {
        return new Effect(modelName, a, b);
    }

    /**
     * 设置特效大小
     */
    public set effectSize(size: number) {
        EXSetEffectSize(this.handle, size)
        this.data.size = size
    }

    /**
     * 获取特效大小
     */
    public get effectSize() {
        let size = this.data.size
        if (!size) {
            size = EXGetEffectSize(this.handle)
        }
        return size
    }

    /**
     * 设置特效粒子大小
     */
    public set effectPariticleSize(size: number) {
        if (SetPariticle2Size) {
            SetPariticle2Size(this.handle, size)
        }
        this.data.Parsize = size
    }

    /**
     * 获取特效粒子大小
     */
    public get effectPariticleSize() {
        let size = this.data.Parsize
        if (!size) {
            size = 0
        }
        return size
    }

    /**
     * 设置特效高度
     */
    public set effectZ(Z: number) {
        EXSetEffectZ(this.handle, Z)
        this.data.z = Z
    }

    /**
     * 获取特效高度
     */
    public get effectZ() {
        let z = this.data.z
        if (!z) {
            z = EXGetEffectZ(this.handle)
        }
        return z
    }

    /**
     * 设置特效播放速度
     */
    public set effectSpeed(Speed: number) {
        EXSetEffectSpeed(this.handle, Speed)
        this.data.Speed = Speed
    }

    /**
     * 获取特效播放速度
     */
    public get effectSpeed() {
        return this.data.Speed
    }

    /**
     * 设置特效颜色
     */
    public set effectColor(Color: number) {
        EXSetEffectColor(this.handle, Color)
        this.data.color = Color
    }

    /**
     * 获取特效颜色
     */
    public get effectColor() {
        let Color = this.data.color
        if (!Color) {
            Color = EXGetEffectColor(this.handle)
        }
        return Color
    }

    /**
     * 设置迷雾可见特效
     */
    public set effectFogVisible(fogvisible: boolean) {
        EXSetEffectFogVisible(this.handle, fogvisible)
        this.data.fogvisible = fogvisible
    }

    /**
     * 获取特效迷雾可见情况
     */
    public get effectFogVisible() {
        return this.data.fogvisible
    }

    /**
     * 设置特效可见
     */
    public set effectVisible(visible: boolean) {
        EXSetEffectVisible(this.handle, visible)
        this.data.visible = visible
    }

    /**
     * 设置特效可见
     */
    public get effectVisible() {
        return this.data.visible
    }

    /**
     * 设置特效Z轴旋转
     */
    public set effectRotateZ(z: number) {
        this.reset()
        EXEffectMatRotateZ(this.handle, z)
        this.data.rotateZ = z
    }

    /**
     * 获取特效Z轴旋转
     */
    public get effectRotateZ() {
        return this.data.rotateZ
    }

    /**
     * 设置特效X轴旋转
     */
    public set effectRotateX(x: number) {
        EXEffectMatRotateX(this.handle, x)
        this.data.rotateX = x
    }

    /**
     * 获取特效X轴旋转
     */
    public get effectRotateX() {
        return this.data.rotateX
    }

    /**
     * 设置特效Y轴旋转
     */
    public set effectRotateY(y: number) {
        EXEffectMatRotateY(this.handle, y)
        this.data.rotateY = y
    }

    /**
     * 获取特效Y轴旋转
     */
    public get effectRotateY() {
        return this.data.rotateY
    }

    /**
     * 移动特效到XY
     */
    public setEffectXY(x: number, y: number) {
        EXSetEffectXY(this.handle, x, y)
        this.data.x = x
        this.data.y = y
    }

    /**
     * 获取特效XY轴
     */
    public getEffectXY(): { x: number, y: number } {
        let xy = {x: this.data.x, y: this.data.y}
        if (!xy.x) {
            xy.x = EXGetEffectX(this.handle)
        }
        if (!xy.y) {
            xy.y = EXGetEffectY(this.handle)
        }
        return xy
    }

    /**
     * 设置特效绑定handle
     */
    public setBindEffect(han: handle, node: string) {
        BindEffect(han, node, this.handle)
        this.data.binding = han
    }

    /**
     * 获取特效绑定的handle
     */
    public getBindEffect() {
        return this.data.binding
    }

    /**
     * 解除特效绑定
     */
    public unBindEffect() {
        UnBindEffect(this.handle)
        this.data.binding = null
    }

    /**
     * 设置特效播放动画
     */
    public setEffectAnimation(aon: number) {
        EXSetEffectAnimation(this.handle, aon)
    }

    /**
     * 特效重置（清空特效旋转与缩放）
     */
    public reset() {
        EXEffectMatReset(this.handle)
        this.data.size = 1
        this.data.rotateX = 0
        this.data.rotateY = 0
        this.data.rotateZ = 0
    }

    /**
     * 摧毁特效
     */
    public destroy() {
        DestroyEffect(this.handle);
    }

    public static fromHandle(handle: effect): Effect {
        return this.getObject(handle);
    }
}

File: scripts/solar/solar-common/w3ts/handles/effect.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__New = ____lualib.__TS__New
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["9"] = 3,["10"] = 3,["11"] = 6,["12"] = 6,["13"] = 6,["14"] = 6,["15"] = 24,["16"] = 25,["17"] = 6,["18"] = 27,["19"] = 6,["20"] = 28,["21"] = 28,["22"] = 28,["23"] = 29,["24"] = 6,["25"] = 30,["26"] = 30,["27"] = 30,["29"] = 24,["30"] = 154,["31"] = 155,["32"] = 154,["33"] = 161,["34"] = 162,["35"] = 161,["36"] = 337,["37"] = 338,["38"] = 339,["39"] = 340,["40"] = 337,["41"] = 346,["42"] = 347,["43"] = 348,["44"] = 349,["46"] = 351,["47"] = 352,["49"] = 354,["50"] = 346,["51"] = 360,["52"] = 361,["53"] = 362,["54"] = 360,["55"] = 368,["56"] = 369,["57"] = 368,["58"] = 375,["59"] = 376,["60"] = 377,["61"] = 375,["62"] = 383,["63"] = 384,["64"] = 383,["65"] = 390,["66"] = 391,["67"] = 392,["68"] = 393,["69"] = 394,["70"] = 395,["71"] = 390,["72"] = 401,["73"] = 402,["74"] = 401,["75"] = 405,["76"] = 406,["77"] = 405,["83"] = 12,["84"] = 13,["86"] = 15,["88"] = 18,["89"] = 19,["99"] = 177,["100"] = 178,["101"] = 179,["103"] = 181,["105"] = 168,["106"] = 169,["107"] = 170,["117"] = 198,["118"] = 199,["119"] = 200,["121"] = 202,["123"] = 187,["124"] = 188,["125"] = 189,["127"] = 191,["137"] = 217,["138"] = 218,["139"] = 219,["141"] = 221,["143"] = 208,["144"] = 209,["145"] = 210,["155"] = 236,["157"] = 227,["158"] = 228,["159"] = 229,["169"] = 251,["170"] = 252,["171"] = 253,["173"] = 255,["175"] = 242,["176"] = 243,["177"] = 244,["187"] = 270,["189"] = 261,["190"] = 262,["191"] = 263,["201"] = 285,["203"] = 276,["204"] = 277,["205"] = 278,["215"] = 301,["217"] = 291,["218"] = 292,["219"] = 293,["220"] = 294,["230"] = 316,["232"] = 307,["233"] = 308,["234"] = 309,["244"] = 331,["246"] = 322,["247"] = 323,["248"] = 324});
local ____exports = {}
local ____handle = require("solar.solar-common.w3ts.handles.handle")
local Handle = ____handle.Handle
____exports.Effect = __TS__Class()
local Effect = ____exports.Effect
Effect.name = "Effect"
__TS__ClassExtends(Effect, Handle)
function Effect.prototype.____constructor(self, modelName, a, b)
    if Handle:initFromHandle() then
        Handle.prototype.____constructor(self)
    elseif type(a) == "number" and type(b) == "number" then
        Handle.prototype.____constructor(
            self,
            AddSpecialEffect(modelName, a, b)
        )
    elseif type(a) ~= "number" and type(b) == "string" then
        Handle.prototype.____constructor(
            self,
            AddSpecialEffectTarget(modelName, a.handle, b)
        )
    end
end
function Effect.createEffect(self, modelName, x, y)
    return __TS__New(____exports.Effect, modelName, x, y)
end
function Effect.createEffectTarget(self, modelName, a, b)
    return __TS__New(____exports.Effect, modelName, a, b)
end
function Effect.prototype.setEffectXY(self, x, y)
    EXSetEffectXY(self.handle, x, y)
    self.data.x = x
    self.data.y = y
end
function Effect.prototype.getEffectXY(self)
    local xy = {x = self.data.x, y = self.data.y}
    if not xy.x then
        xy.x = EXGetEffectX(self.handle)
    end
    if not xy.y then
        xy.y = EXGetEffectY(self.handle)
    end
    return xy
end
function Effect.prototype.setBindEffect(self, han, node)
    BindEffect(han, node, self.handle)
    self.data.binding = han
end
function Effect.prototype.getBindEffect(self)
    return self.data.binding
end
function Effect.prototype.unBindEffect(self)
    UnBindEffect(self.handle)
    self.data.binding = nil
end
function Effect.prototype.setEffectAnimation(self, aon)
    EXSetEffectAnimation(self.handle, aon)
end
function Effect.prototype.reset(self)
    EXEffectMatReset(self.handle)
    self.data.size = 1
    self.data.rotateX = 0
    self.data.rotateY = 0
    self.data.rotateZ = 0
end
function Effect.prototype.destroy(self)
    DestroyEffect(self.handle)
end
function Effect.fromHandle(self, handle)
    return self:getObject(handle)
end
__TS__SetDescriptor(
    Effect.prototype,
    "data",
    {
        get = function(self)
            if not self._data then
                self._data = {}
            end
            return self._data
        end,
        set = function(self, value)
            self._data = value
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectSize",
    {
        get = function(self)
            local size = self.data.size
            if not size then
                size = EXGetEffectSize(self.handle)
            end
            return size
        end,
        set = function(self, size)
            EXSetEffectSize(self.handle, size)
            self.data.size = size
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectPariticleSize",
    {
        get = function(self)
            local size = self.data.Parsize
            if not size then
                size = 0
            end
            return size
        end,
        set = function(self, size)
            if SetPariticle2Size then
                SetPariticle2Size(self.handle, size)
            end
            self.data.Parsize = size
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectZ",
    {
        get = function(self)
            local z = self.data.z
            if not z then
                z = EXGetEffectZ(self.handle)
            end
            return z
        end,
        set = function(self, Z)
            EXSetEffectZ(self.handle, Z)
            self.data.z = Z
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectSpeed",
    {
        get = function(self)
            return self.data.Speed
        end,
        set = function(self, Speed)
            EXSetEffectSpeed(self.handle, Speed)
            self.data.Speed = Speed
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectColor",
    {
        get = function(self)
            local Color = self.data.color
            if not Color then
                Color = EXGetEffectColor(self.handle)
            end
            return Color
        end,
        set = function(self, Color)
            EXSetEffectColor(self.handle, Color)
            self.data.color = Color
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectFogVisible",
    {
        get = function(self)
            return self.data.fogvisible
        end,
        set = function(self, fogvisible)
            EXSetEffectFogVisible(self.handle, fogvisible)
            self.data.fogvisible = fogvisible
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectVisible",
    {
        get = function(self)
            return self.data.visible
        end,
        set = function(self, visible)
            EXSetEffectVisible(self.handle, visible)
            self.data.visible = visible
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectRotateZ",
    {
        get = function(self)
            return self.data.rotateZ
        end,
        set = function(self, z)
            self:reset()
            EXEffectMatRotateZ(self.handle, z)
            self.data.rotateZ = z
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectRotateX",
    {
        get = function(self)
            return self.data.rotateX
        end,
        set = function(self, x)
            EXEffectMatRotateX(self.handle, x)
            self.data.rotateX = x
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectRotateY",
    {
        get = function(self)
            return self.data.rotateY
        end,
        set = function(self, y)
            EXEffectMatRotateY(self.handle, y)
            self.data.rotateY = y
        end
    },
    true
)
return ____exports
